src/template/Template.js

```diff
@@ -34,96 +33,83 @@ export const Template = class {
-    const copy = { ...context };
+  async renderMulti(context, flexField, ai, cache, timer) {
+    const maxTokens = ai.maxTokens || 128000;
+
+    let key;
+    if (cache) {
+      const hash = hashObjectShort({
+        context,
+        flexField,
+        maxTokens,
+        base: this.base,
+      });
+      key = `template-renderMulti-${this.constructor.name}-${hash}`;
+      const cached = await cache.get(key);
+      if (cached) return cached;
+    }
+
+    timer?.log('start');
+
+    const barePrompt = this.render({ ...context, [flexField]: '' });
+    const barePromptTokens = await ai.countTokens(barePrompt);
+
+    timer?.log(`count bare prompt tokens, got ${barePromptTokens}`);
+
+    const remainingTokens = (ai.maxTokens || 128000) - barePromptTokens;
+
     const prompts = [];
-    const offset = 0;
-    while (true) {
-      const { prompt, bytesUsed, done } = await this.renderCapped(copy, flexField, ai, cache);
+    for (let i = 0; i < context[flexField].length; i += remainingTokens) {
+      const prompt = this.render({
+        ...context,
+        [flexField]: context[flexField].slice(i, i + remainingTokens),
+      });
       prompts.push(prompt);
-      if (done) {
-        break;
-      }
-      copy[flexField] = copy[flexField].substr(bytesUsed);
     }
+
+    if (cache) cache.set(key, prompts);
+
     return prompts;
   }
 
   async renderCapped(context, flexField, ai, cache) {
     const maxTokens = ai.maxTokens || 128000;
-    const countFn = async (str) => ai.countTokens(str);
-    const accuracy = 8000;
 
     let key;
     if (cache) {
-      const hash = CryptoJS.SHA256(
-        JSON.stringify({
-          context,
-          accuracy,
-          flexField,
-          maxTokens,
-          base: this.base,
-        }),
-      )
-        .toString(CryptoJS.enc.Hex)
-        .substr(0, 16);
+      const hash = hashObjectShort({
+        context,
+        flexField,
+        maxTokens,
+        base: this.base,
+      });
       key = `template-renderCapped-${this.constructor.name}-${hash}`;
       const cached = await cache.get(key);
-      if (cached) {
-        return cached;
-      }
+      if (cached) return cached;
     }
 
-    const start = new Date().getTime();
+    const barePrompt = this.render({ ...context, [flexField]: '' });
+    const barePromptTokens = await ai.countTokens(barePrompt);
+    const remainingTokens = (ai.maxTokens || 128000) - barePromptTokens;
 
-    let prompt;
-    let tokens;
-    let guess = Math.min(context[flexField].length, maxTokens * 4);
-    let lowerBound = 0;
-    let upperBound = Math.min(context[flexField].length, maxTokens * 8);
+    const flexContent = context[flexField];
 
-    const render = (size) => {
-      const copy = { ...context };
-      copy[flexField] = context[flexField].substr(0, size);
-      return this.render(copy);
+    const result = {
+      prompt: this.render({
+        ...context,
+        [flexField]: context[flexField].slice(0, remainingTokens),
```

Comment:
per convo, should modify this to get close to the max token limit, doing chunks of 1024 bytes sounds reasonable